### ISSUE TITLE
[API Proposal]: 

### ISSUE THREAD

**Issue by polijp** [dotnet/runtime#64187]
### Background and motivation

.net 6 introduced CollectionsMarshall to access a List elements as a Span or to access to a value of a dictionary with a ref struct.
These features are really useful in mathematical computing that needs high performances.

The idea here consists in expanding CollectionsMarshall with SortedList that maintains two internal arrays (one for the keys, one for the values). To guarantee the sorted aspect, one will gets a ReadOnlySpan. It will allow to access the keys and/or the values and browse them quicker than with the Keys and Values Properties.

### API Proposal

```C#
namespace System.Runtime.InteropServices
{
    public static class CollectionsMarshal
    {
        public ReadOnlySpan<TKey> GetKeysAsReadOnlySpan<TKey, TValue>(SortedList<TKey,TValue> collection);
        public ReadOnlySpan<TValue> GetValuesAsReadOnlySpan<TKey, TValue>(SortedList<TKey,TValue> collection);
    }
}
```

### API Usage

```C#
// Browse the keys/values
var span = CollectionsMarshal.GetKeysAsReadOnlySpan(mycollection);
for(int i=0; i<span.Length;i++)
  Console.WriteLine(span[i]);

// Sums the values
SortedList<string, int> mycollection = new();
...
var values = CollectionsMarshal.GetValuesAsReadOnlySpan(mycollection);
int vec_count = Vector<int>.Count;
int sum = 0;

int i = 0;

// vectorized sum
for (; i <= values.Length - vec_count; i += vec_count)
    sum += Vector.Sum(new Vector<int>(values.Slice(i));

for (; i<values.Length ; i++)
    sum += values[i];

Console.WriteLine(sum);
```

### Risks

There is a risk when the collection is changed, either we have to refresh the length of the ReadOnlySpan that points on the keys (or the values), either we decide it keeps the length of the collection when it was created.

**Comment by Joe4evr** [comment]
> There is a risk when the collection is changed, either we have to refresh the length of the ReadOnlySpan that points on the keys (or the values), either we decide it keeps the length of the collection when it was created.

It's not possible to change the length of a span instance after construction (just like arrays). Also, the same risk of contents changing already exists with `AsSpan<T>(List<T>)`, you simply have to know what you're doing.

**Comment by kasperk81** [comment]
> [API Proposal]:

missing the title?

**Comment by KalleOlaviNiemitalo** [comment]
> To guarantee the sorted aspect, one will gets a ReadOnlySpan.

The span of values wouldn't have to be read-only because the values won't affect sorting.

**Comment by stephentoub** [comment]
Note this wouldn't just be adding a new API... it would require moving `SortedList<>` from System.Collections.dll to System.Private.CoreLib.dll.

**Comment by polijp** [comment]
If moving SortedList<> is a big problem, can't we use [InternalsVisibleTo] and change the fields to internal?

**Comment by stephentoub** [comment]
>  can't we use [InternalsVisibleTo] and change the fields to internal

No, even if we were ok using InternalsVisibleTo (which we avoid), corelib can't reference collections.dll, it's the other way around.

**Comment by stephentoub** [comment]
Also, once we stop caring about BinaryFormatter and thus the impact of changes on serialization compat, we could very likely be interested in changing how data is stored in SortedList, but a method like this adds back such constraints.

I think such a method would need to yield very significant benefits to consider adding it, given what it costs in flexibility.

**Comment by EgorBo** [comment]
> we could very likely be interested in changing how data is stored in SortedList, but a method like this adds back such constraints.

I guess it was ok for `CollectionsMarshal.AsSpan` for list which we would never change internally to some other data structure, but for dictionaries/sortinglist it may happen, e.g. some tree-based

**Comment by dusrdev** [comment]
I think the need for this is significant, currently SortedList is great for CRUD operations because of the inherent order. But sometimes the order can be leveraged to the entire collection rather than a single item.

For example, let's say I have `SortedList<DateTime, Order>`, more than just doing CRUD on a single `Order`, if I would have access to these APIs I would then be able to:

```c#
SortedList<DateTime, Order> lst = []; // imagine contains orders
Order order = .... // some order
var index = lst.IndexOfKey(order.Date);
ReadOnlySpan<Order> span = CollectionsMarshal.GetValuesAsReadOnlySpan(lst);
ReadOnlySpan<Order> olderThanReference = span.Slice(0, index);
ReadOnlySpan<Order> newerThanReference = span.Slice(index);
```

Which allows me to iterate these parts of the collection very efficiently.

[#99448](https://github.com/dotnet/runtime/issues/99448) Makes this even more useful, as an arbitrary date that is not part of an existing Order could be used as reference.

This is so useful that in-fact, In one of my projects I am doing this currently with List<Order> but than I am forced to create a comparer, and enforce it for sorting, and manually facilitate for all CRUD using the BinarySearch function. But this is error-prone.
